core/src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20210615_sharded_entity_tables.xml

```diff
@@ -0,0 +1,271 @@
+                    ' CONSTRAINT FK_ENT_ATTRIBUTE_ENTITY_REF_',@tableSuffix,' FOREIGN KEY (value_entity_ref) REFERENCES ENTITY (id)'
+                    ') ENGINE=InnoDB AUTO_INCREMENT=231805857 DEFAULT CHARSET=utf8;');
+                PREPARE stmt from @sql_text;
+                EXECUTE stmt;
+            END /
+
+            -- procedure to copy rows from ENTITY_ATTRIBUTE to the appropriate shard table,
+            -- based on the attributes' owning workspace
+            DROP PROCEDURE IF EXISTS copyRowsToAttributeShard /
+            CREATE PROCEDURE copyRowsToAttributeShard(IN shardIdentifier CHAR(4))
+            BEGIN
+                SET @tableSuffix := shardIdentifier;
+                SET @sql_text:=CONCAT('INSERT INTO ENTITY_ATTRIBUTE_',@tableSuffix,'(id, name, value_string, value_number, value_boolean,'
+                    ' value_entity_ref, list_index, owner_id, list_length, namespace, VALUE_JSON, deleted, deleted_date)'
+                    'SELECT ea.id, ea.name, ea.value_string, ea.value_number, ea.value_boolean,'
+                    ' ea.value_entity_ref, ea.list_index, ea.owner_id, ea.list_length, ea.namespace, ea.VALUE_JSON, ea.deleted, ea.deleted_date'
+                    ' FROM ENTITY_ATTRIBUTE ea, ENTITY e, WORKSPACE w'
+                    ' WHERE e.workspace_id = w.id'
+                    ' AND ea.owner_id = e.id'
+                    ' AND shardIdentifier(hex(w.id)) = ''',@tableSuffix,''';');
+                PREPARE stmt from @sql_text;
+                EXECUTE stmt;
+            END /
+
+            -- convenience procedure to both create a shard and copy rows into it
+            DROP PROCEDURE IF EXISTS createAndPopulateShard /
+            CREATE PROCEDURE createAndPopulateShard(IN shardIdentifier CHAR(4))
+            BEGIN
+                CALL createEntityAttributeShard(shardIdentifier);
+                CALL copyRowsToAttributeShard(shardIdentifier);
+            END /
+        </sql>
+        <rollback>
+            DROP PROCEDURE IF EXISTS copyRowsToAttributeShard;
+            DROP PROCEDURE IF EXISTS createEntityAttributeShard;
+            DROP FUNCTION IF EXISTS shardIdentifier;
+        </rollback>
+    </changeSet>
+
+    <!-- START create and populate each shard in its own changeset. If we hit an error with an
+            individual shard, we can restart the liquibase migration at that point,
+            and don't have to rollback any progress up to that changeset. -->
+    <changeSet logicalFilePath="dummy" author="davidan" id="create_and_populate_shard_0_07">
+        <sql>call createAndPopulateShard('0_07');</sql>
+        <rollback>DROP TABLE IF EXISTS ENTITY_ATTRIBUTE_0_07;</rollback>
+    </changeSet>
+
+    <changeSet logicalFilePath="dummy" author="davidan" id="create_and_populate_shard_0_8f">
+        <sql>call createAndPopulateShard('0_8f');</sql>
+        <rollback>DROP TABLE IF EXISTS ENTITY_ATTRIBUTE_0_8f;</rollback>
+    </changeSet>
```

Comment:
the copy-paste in this file is super ugly and as a programmer I hate it. But I think it's the right call to have an individual liquibase changeset for each shard, and I don't know how else to do this via liquibase.

Reply:
It's awesome to have all of your database changes in source control alongside the application's code, but I do feel this change is so big that we're going to run into problems trying to use liquibase here.  But, if we can use liquibase, then this seems like the right approach.

Reply:
I think there is a way to implement new Liquibase tags in Scala or Java that do whatever you want. Not sure if that's any better than this for something one-off, though

Reply:
it looks like the custom tags allow you to implement custom logic within a changeset … what I really want is to be able to programmatically generate changesets. (what I did for this one is use Scastie to loop over the shard ids and output the changesets, which I then cut and paste into the file)